src/passes/AggressiveMemoryOffsetMerging.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Optimize load and store instructions aggressively.
+// This pass assumes that out of bound memory usage is protected in high level
+// language grammar. In WASM bytecode level, we can treat offset in memory
+// load and store as normal arithmetic operations without overflow.
+//
+//   load offset=A
+//     i32.add
+//       i32.const B
+//       other expr
+// Then this pattern can be optimized to
+//   load offset=0
+//     i32.add
+//       i32.const A + B
+//       other expr
```

Comment:
I believe the existing OptimizeAddedConstants pass can do something like this, using the assumption of `--low-memory-unused` (it will also fold `A + B` into the offset, saving the add).

Does that achieve what you want in this pass?

Reply:
You can test it with `wasm-opt --optimize-added-constants --low-memory-unused`. Or `wasm-opt -O3 --low-memory-unused` (that pass is enabled when that flag is set).

Reply:
This direction of these two passes are different. This pass want to merge offset to add/sub const. It will not reduce the code size in most of cases. Just simplify the execution of load / store instructions for the interpret / baseline compiler. `--optimize-added-constants --low-memory-unused` actually can provide better optimization because it will merge add/sub const into offset.
For this pass, `low-memory-unused` is not enough, it needs more loose assumption which memory visit should never out of bound.

Reply:
The usecase is in AssemblyScript which is a language has internal GC. And the GC info is stored before the object. It will produce the code `load(pos=&ptr[-20], offset=16)`. and it can be simplify a bit in this pass.
<img width="1257" alt="image" src="https://github.com/user-attachments/assets/6eb0268f-ece9-4bc0-8073-7f4e6ccf868e">

Reply:
Thanks for the example. Ok, so IIUC you want to optimize
```wat
    (i32.load offset=16
      (i32.sub
        (local.get $x)
        (i32.const 20)
      )
    )
```
into
```wat
    (i32.load
      (i32.sub
        (local.get $x)
        (i32.const 4)
      )
    )
```
Unless I'm missing something, I think that could be optimized under the assumption of low memory unused. We don't do it atm, looks like we only handle adds, but subtracts could be added. The reason it is valid to optimize is that the first expression is really

```c
((x - 20) & 0xffffff) + 16
```
(`i32.sub` can overflow/underflow, which the `&` handles; `offset=16` does not overflow so no fixup is needed there). The question is then whether that is equal to
```c
(x - 4) & 0xffffff
```
By the [assumption of low memory unused](https://github.com/WebAssembly/binaryen/blob/fcd8dfe7ddd1a4a3ce1be40c674fdd60bbec73d1/src/pass.h#L174-L178), `x >= 1024`. That means the `&` operations in both cases never do any work. So `x - 20 + 16 === x - 4`, and optimizing is valid.

Maybe I got something wrong though, what do you think @HerrCai0907 ?

Reply:
Actually, I did get something wrong, the assumption is that the load pointer is `>= 1024`, not `x`. That is,
```c
((x - 20) & 0xffffff) + 16 >= 1024
```
But I don't think that changes the conclusion.

Reply:
That is true. I agree that we may can merge this kind of optimization to `optimize-added-constants`.

But I think these two assumptions are still different. For AS (which implement allocator and GC in language level). I can easily assume no out of bound memory access but hard to say no low memory access. For C++, it is the same, since out of bound memory access is UB, but low offset memory access is not.

Do you think we should add this flag also to enable this kind of optimization also for this assumption?

Reply:
I agree that AS might have even stronger assumptions, because of what you said. Perhaps those assumptions allow more to be optimized. But for this specific optimization, since low-memory-unused is enough for it, I think we should implement it in optimized-added-constants, and without adding a new flag.

If we find other optimizations later that require strong assumptions that AS supports, then I am not opposed to adding more things then, of course.

Reply:
I think it again. If binaryen's approach to traps in optimization is to treat them as UB, then the optimization does not even require this assumption. Because it will change result iff overflow happen during cal offset (trap -> visit incorrect address).

Reply:
Binaryen doesn't treat traps as UB, or at least not in sense that LLVM might. By default we do not remove traps at all, though we do not care about the identity of traps (so we can reorder two traps). We do have options to do more, and ignore/remove certain traps:

https://github.com/WebAssembly/binaryen/blob/bc36c02d1a54c91d9fc4bdbffe00608929ec3169/src/pass.h#L122-L173

Reply:
Then I think we should put this optimization under `trapsNeverHappen` options.
Back to the example
```
(i32.load offset=16 (i32.sub (local.get $x) (i32.const 20)))
```
and
```
(i32.load offset=0  (i32.sub (local.get $x) (i32.const 4)))
```

if x is 16, in original code, it will visit 0x1'0000'0000C but optimized code will visit 0xC.

Reply:
I would agree if we didn't have low-memory-unused, which I thought we agreed above is correct for this? We don't need both flags here, as low-memory-unused defines exactly this situation.